README.md

```diff
@@ -1,526 +1,917 @@
+
+  ```js
+  class Example {
+    @reactive prop myBool = false;
   }
+  ```
 
-  @logged
-  set #x(value) { }
-}
+- Class _initialized methods_, defined by applying the `init` keyword to a class method. These methods are defined on the prototype of the class, but then assigned to a slot with the same name on the instance:
 
-new C().m(1);
-// starting m with arguments 1
-// starting set #x with arguments 1
-// ending set #x
-// ending m
+  ```js
+  class Example {
+    @bound init myMethod() {
+      // ...
+    }
+  }
+  ```
+
+These new element types can be used independently, and have their own semantics separate from usage with decorators. The reason they are included in this proposal is primarily because there are a number of use cases for decorators which require their semantics, since decorators can only replace an element with a corresponding element that has the same semantics. These use cases are common in the existing decorators ecosystem, demonstrating a need for the capabilities they provide.
+
+## Detailed Design
+
+In general, decorators receive two parameters:
+
+1. The value being decorated, or `undefined` in the case of class fields which are a special case (see below for details).
+2. A context object containing metadata about the value being decorated
+
+Using TypeScript interfaces for brevity and clarity, this is the general shape of the API:
+
+```ts
+type Decorator = (value: Input, context: {
+  kind: string;
+  name?: string;
+  access?: {
+    get?(): unknown;
+    set?(value: unknown): void;
+  };
+  isPrivate?: boolean;
+  isStatic?: boolean;
+  defineMetadata(key: string | symbol | number, value: unknown);
+}) => Output | void;
 ```
 
-`@logged` can be implemented in JavaScript as a decorator. Decorators are functions that are called with an argument containing what's being decorated. For example:
-- A method decorator is called with the method being decorated
-- A getter decorator is called with the getter function being decorated
-- A setter decorator is called with the setter function being decorated
+`Input` and `Output` here represent the values passed to and returned from a given decorator. Each type of decorator has a different input and output, and these are covered below in more detail. All decorators can choose to return nothing, which defaults to using the original, undecorated value.
+
+The context object also varies depending on the value being decorated. Breaking down the properties:
+
+- `kind`: The kind of decorated value. This can be used to assert that the decorator is used correctly, or to have different behavior for different types of values. It is one of the following values.
+  - `"class"`
+  - `"method"`
+  - `"init-method"`
+  - `"getter"`
+  - `"setter"`
+  - `"field"`
+  - `"prop"`
```

Comment:
Given the discussion below, I'm led to wonder whether there's value in distinguishing private and public here as well—given that there are values you can *count on* being present or absent for public or private fields or methods respectively. 🤔

Reply:
Yeah, that makes sense. 👍🏼